### PR TITLE
Added support for infoblox member object

### PR DESCRIPTION
**Pull request by krisvasudevan**
##### SUMMARY

Added support for member object. It supports adding and/or removes Infoblox NIOS servers.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

nios_member

##### ADDITIONAL INFORMATION

The module supports various parameters that enable adding a standalone/HA member. It also supports both IPv4 and IPv6.
Changes had to be made to api.py file in order to normalize the input parameters. The member_normalize function will remove any arguments that are set to None since WAPI will error on that condition.

```paste below

```

### CODE REVIEW

File: lib/ansible/module_utils/net_tools/nios/api.py [hunk 1]
```diff
@@ -31,7 +31,6 @@
 from ansible.module_utils._text import to_native
 from ansible.module_utils.six import iteritems
 from ansible.module_utils._text import to_text
-from ansible.module_utils.basic import env_fallback
```
Comment (justjais) [inline lib/ansible/module_utils/net_tools/nios/api.py]: @krisvasudevan why have you removed this line?

File: lib/ansible/module_utils/net_tools/nios/api.py [hunk 2]
```diff
@@ -55,25 +54,20 @@
 NIOS_SRV_RECORD = 'record:srv'
 NIOS_NAPTR_RECORD = 'record:naptr'
 NIOS_TXT_RECORD = 'record:txt'
-NIOS_NSGROUP = 'nsgroup'
```
Comment (justjais) [inline lib/ansible/module_utils/net_tools/nios/api.py]: Seems like your branch is not updated, else why have you removed the support for the existing modules.

File: lib/ansible/module_utils/net_tools/nios/api.py [hunk 3]
```diff
@@ -55,25 +54,20 @@
...
-NIOS_IPV4_FIXED_ADDRESS = 'fixedaddress'
-NIOS_IPV6_FIXED_ADDRESS = 'ipv6fixedaddress'
-NIOS_NEXT_AVAILABLE_IP = 'func:nextavailableip'
-NIOS_IPV4_NETWORK_CONTAINER = 'networkcontainer'
-NIOS_IPV6_NETWORK_CONTAINER = 'ipv6networkcontainer'
+NIOS_MEMBER = 'member'
 
 NIOS_PROVIDER_SPEC = {
-    'host': dict(fallback=(env_fallback, ['INFOBLOX_HOST'])),
```
Comment (justjais) [inline lib/ansible/module_utils/net_tools/nios/api.py]: `env_fallback` module is used when user has defined `INFOBLOX_HOST` value under env variable. Please undo these changes.

File: lib/ansible/module_utils/net_tools/nios/api.py [hunk 4]
```diff
@@ -114,8 +108,8 @@ def normalize_extattrs(value):
         extattrs: {
             key: {
                 value: <value>
-            }
-        }
+                }
```
Comment (justjais) [inline lib/ansible/module_utils/net_tools/nios/api.py]: Why the formatting?

File: lib/ansible/module_utils/net_tools/nios/api.py [hunk 5]
```diff
@@ -222,29 +238,24 @@ def run(self, ib_obj_type, ib_spec):
         # get object reference
         ib_obj_ref, update, new_name = self.get_object_ref(self.module, ib_obj_type, obj_filter, ib_spec)
 
-        proposed_object = {}
```
Comment (justjais) [inline lib/ansible/module_utils/net_tools/nios/api.py]: There's seems to be many changes that I am unable to understand, from this down below can u comment that why you have made all the changes. Thanks

File: lib/ansible/modules/net_tools/nios/nios_member.py [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'certified'}
+
+DOCUMENTATION = '''
+---
+module: nios_member
```
Comment (justjais) [inline lib/ansible/modules/net_tools/nios/nios_member.py:14]: Please write the unit tests for this module.

**Review by justjais — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/net_tools/nios/nios_member.py [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+          hwmodel:
+            description:
+              - Hardware model
+          hwtype:
+            description:
+              - Hardware type.
+      licenses:
+        description:
+          - An array of license types the pre-provisioned member should have in order to join the Grid, or the licenses that must be allocated to the member when it joins the Grid.
+    description:
```
Comment (justjais) [inline lib/ansible/modules/net_tools/nios/nios_member.py:287]: State keyword is missing.

File: lib/ansible/module_utils/net_tools/nios/api.py [hunk 8]
```diff
@@ -432,4 +460,3 @@ def on_update(self, proposed_object, ib_spec):
             update = ib_spec[key].get('update', True)
             if not update:
                 keys.add(key)
-        return dict([(k, v) for k, v in iteritems(proposed_object) if k not in keys])
```
Comment (justjais) [inline lib/ansible/module_utils/net_tools/nios/api.py]: Kindly undo this change as this returns value from `on_update` function.

File: lib/ansible/module_utils/net_tools/nios/api.py [hunk 9]
```diff
@@ -136,6 +137,29 @@ def flatten_extattrs(value):
     return dict([(k, v['value']) for k, v in iteritems(value)])
 
 
+def member_normalize(member_spec):
+    ''' Transforms the member module arguments into a valid WAPI struct
```
Comment (justjais) [inline lib/ansible/module_utils/net_tools/nios/api.py:141]: Can you give me an example which this recursive function is trying to resolve, coz this is heavy operation and need to check if your use case can be resolved in other ways.

**Comment by justjais** [comment]
@krisvasudevan please try to fix all the shippable errors.

File: lib/ansible/module_utils/net_tools/nios/api.py (hunk 9 above)
Comment (krisvasudevan) [reply lib/ansible/module_utils/net_tools/nios/api.py:141]: @justjais 
When no values are passed to some of the options in the module, it takes the value "None" like below.
`{
	'comment': 'CreatedbyAnsible',
	'enable_ha': True,
	'node_info': [
		{
			'v6_mgmt_network_setting': None,
			'lan_ha_port_setting': {
				'mgmt_ipv6addr': None,
				'ha_port_setting': None,
				'lan_port_setting': None,
				'ha_ip_address': '192.168.1.70',
				'mgmt_lan': '192.168.1.80'
			},
			'mgmt_network_setting': None,
			'lan2_physical_setting': None
		},
		{
			'v6_mgmt_network_setting': None,
			'lan_ha_port_setting': {
				'mgmt_ipv6addr': None,
				'ha_port_setting': None,
				'lan_port_setting': None,
				'ha_ip_address': '192.168.1.71',
				'mgmt_lan': '192.168.1.81'
			},
			'mgmt_network_setting': None,
			'lan2_physical_setting': None
		}
	],
	'config_addr_type': 'IPV4',
	'router_id': 150,
	'host_name': 'member-ha.ansible-dev.com',
	'platform': 'VNIOS',
	'vip_setting': {
		'subnet_mask': '255.255.255.0',
		'gateway': '192.168.1.1',
		'address': '192.168.1.100'
	},
	'upgrade_group': 'Default'
}`
The WAPI errors out in this situation with an error that requests for the right kind of input.
In order to avoid this, this function strips all the options set to to None from input , like below:
`{
	'comment': 'CreatedbyAnsible',
	'enable_ha': True,
	'node_info': [
		{
			'lan_ha_port_setting': {
				'ha_ip_address': '192.168.1.70',
				'mgmt_lan': '192.168.1.80'
			}
		},
		{
			'lan_ha_port_setting': {
				'ha_ip_address': '192.168.1.71',
				'mgmt_lan': '192.168.1.81'
			}
		}
	],
	'config_addr_type': 'IPV4',
	'router_id': 150,
	'host_name': 'member-ha.ansible-dev.com',
	'platform': 'VNIOS',
	'vip_setting': {
		'subnet_mask': '255.255.255.0',
		'gateway': '192.168.1.1',
		'address': '192.168.1.100'
	},
	'upgrade_group': 'Default'
}`

File: lib/ansible/module_utils/net_tools/nios/api.py (hunk 8 above)
Comment (krisvasudevan) [reply lib/ansible/module_utils/net_tools/nios/api.py]: @justjais done. I've added it back. It was unintentional.

**Comment by justjais** [comment]
@krisvasudevan please update the PR with the unit test case, rest of changes looks fine. Once done, I shall approve and merge the PR.

**Review by justjais — APPROVED** [review]
